### ISSUE TITLE
ssh completion: hostname in known_hosts ignored if alternative hostname has custom port

### ISSUE THREAD

**Issue by ThomasAH** [fish-shell/fish-shell#4374]
fish 2.6.0 from your OBS on Debian jessie

I have /etc/ssh/ssh_known_hosts with multiple hostnames in one line.
Some are regular names like "foo.example.com", others for the same host contain a port number like "[foo.example.com]:443", so the full line looks like
`foo.example.com,[foo.example.com]:443,192.0.2.1,[192.0.2.1]:443 ssh-ed25519 AAAA...`
("numercial" IPv6 IPs work fine, so I did not include them in above example)

The problem is that the completion code tries to ignore entries with custom ports, and if it finds them it ignores the whole line, including the regular hostname.

**Comment by faho** [comment]
As far as I can tell, this comes down to this line in `__fish_print_hostnames`:

```fish
string replace -ra '(\S+) .*' '$1' <$file | string match -r '^[^#|[=]+$' | string split ","
```

The `string replace` removes everything after the first space, the `string match` removes every line starting with `#`, `|` `[` or `=`, and the `split` splits it.

Intuitively, doing the split first would fix this issue, but it would also create false-positives out of comments that include one or more 

At this point, we should start actually consulting the documentation. From [the section on it in the sshd man page](https://man.openbsd.org/sshd#SSH_KNOWN_HOSTS_FILE_FORMAT):

> Lines starting with ‘#’ and empty lines are ignored as comments. 

Which seems to me like in-line comments are not allowed, so we could simply ignore any line starting with a "#" (and possibly whitespace before it? Or any line with a "#" in it?).

Also, we should ignore `@revoked` lines and possibly `@cert-authority` ones (though there we could maybe also offer the non-wildcarded part?), and negated hosts.

So, something like the following might be better:

```fish
# Ignore comments and @-marked lines, remove everything after the first space,
# split on ",", ignore negated/wildcarded hosts and extract hosts with custom port
string match -r '^[^#@]+$' <$file | string replace -r ' .*' '' | string split "," | string match -rv '[!\*\?]' | string replace -r '\[([^]]+)\]:.*' '$1'
```

This will also extract hosts with a custom port - which IIUC will fail connection if the custom port wasn't specified.

@ThomasAH: Can you try that with $file set to your known_hosts file?

**Comment by ThomasAH** [comment]
Above block works fine for our global and my personal known_hosts files, thank you.

One slight problem is that it increases the wall clock time for a full call of __fish_print_hostnames.fish by about 30%-40%, which just pushed it from a little slower than expected to "huh, it does not react, is it broken?" (best of 3: 0.43s vs 0.33s on Intel Xeon CPU E5-2620 @ 2.10GHz, worse on slightly slower systems). But I assume not everyone has such a long list and finding the relevant host after a slight delay is better than not finding it at all.

The old implementation generated 1304 lines, the new 1409 lines (both have duplicates due to some hosts having two entries, e.g. ssh-rsa and ecdsa-sha2-nistp256 or ssh-ed25519).
For comparison: the bash completion happens almost immediately (and finds 1419 entries, how do I list all completion results in bash without copy&paste after tab-tab?)

By the way: openssl 6.7 accepts the # for comments at the start of the line or after whitespace, but ignores it when e.g. used as an additional hostname, so
`foo.example.com,#bar.example.com ssh-...`
still works for foo.example.com.
So I guess matching up to the first non-whitespace is enough. And the last field can have almost any characters anyway, so ignoring lines with # anywhere would be wrong.

**Comment by faho** [comment]
Okay, try this:

```fish
# Ignore hosts that are hashed, commented or @-marked and strip the key.
string replace -r '[#@| ].*' '' <$file \
# Split on ",".
# Ignore negated/wildcarded hosts.
| string split "," | string match -rv '[!\*\?]' \
# Extract hosts with custom port.
| string replace -r '\[([^]]+)\]:.*' '$1'
```

Note that it's expected that this takes a bit longer simply because it does more. I'm not sure how much, because with my tiny known_hosts it's not measurable.

There are a few places where the ssh completion can be sped up (e.g. they use `string match something | string replace something ''` in a few places, which can be replaced with `string replace -f something ''`), but it seems like this part would need some more regex-foo (to cut down on the number of string calls) or improvements in string, generally in fish or a different tool - the bash completions use `awk`.

**Comment by ThomasAH** [comment]
That is slightly faster than the 2.6.0 version (0.32s wall clock), thank you!

Further speeding up of ssh completion sounds like a separate issue :)

**Comment by faho** [comment]
This was supposed to be closed via #4378.
